keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,919 @@
+    Validation expressions must always evaluate a boolean. Token `claims` (payload) are passed to CEL expressions as a dynamic map `decls.NewMapType(decls.String, decls.Dyn)`.
+
+    > NOTE: If an expression returns `false` after evaluation, a `401 Unauthorized` error will be returned
+      and the associated `message` will be logged in the API server logs.
+
+    Example validation rules:
+
+    ```yaml
+    claimValidationRules:
+
+    - expression: 'claims.aud == "charmander" || claims.aud == "bulbasaur"'
+      message: clients other than charmander or bulbasaur are not allowed
+
+    - expression: 'claims.exp - claims.nbf <= 86400'
+      message: total token lifetime must not exceed 24 hours
+    ```
+
+3. `ClaimMappings` - rules to map claims from a token to Kubernetes user info attributes.
+
+    ```go
+    type ClaimMappings struct {
+        // Username represents an option for the username attribute.
+        // Claim must be a singular string claim.
+        // TODO: decide whether to support a distributed claim for username (what are we required to correlate between the data retrieved for distributed claims? sub? something else?). Limit distributed claim support to OIDC things with clientID validation?
+        // Expression must produce a string value.
+        // Possible prefixes based on the config:
+        //     (1) if userName.prefix = "-", no prefix will be added to the username
+        //     (2) if userName.prefix = "" and userName.claim != "email", prefix will be "<issuer.url>#"
+        //     (3) if userName.expression is set instead, result of expression is used as-is without any implicit prefix
+        // (1) and (2) ensure backward compatibility with the --oidc-username-claim and --oidc-username-prefix flags
+        Username PrefixedClaimOrExpression `json:"username"`
+        // Groups represents an option for the groups attribute.
+        // Claim must be a string or string array claim.
+        // Expression must produce a string or string array value.
+        // "", [], missing, and null values are treated as having no groups.
+        // TODO: investigate if you could make a single expression to construct groups from multiple claims. If not, maybe []PrefixedClaimOrExpression?
+        // +optional
+        Groups PrefixedClaimOrExpression `json:"groups,omitempty"`
+        // UID represents an option for the uid attribute.
+        // Claim must be a singular string claim.
+        // Expression must produce a string value.
+        // TODO: this is net new, should it just be expression?
+        // +optional
+        UID ClaimOrExpression `json:"uid,omitempty"`
+        // Extra represents an option for the extra attribute.
+        //
+        // TODO: examples for this?
```

Comment:
use cases ExtraMapping needs to handle today:

1. hard-coded extra key/value

```
key: "foo"
valueExpression: '"bar"'
```

2. hard-coded key, value copying claim value
```
key: "foo"
valueExpression: 'claims.some_claim'
```

3. hard-coded key, value derived from claim value
```
key: "admin"
valueExpression: '(has(claims.is_admin) && claims.is_admin) ? "true":""'
```


possible future way to pull multiple extra values out via expression (confirm cel comprehensions/mapping is powerful enough to transform the input claims into a filtered / transformed `map[string][]string` output):

```
# mutually exclusive with key/valueExpression
keyAndValueExpression: '{"key":"string-value", "key2": ["value1","value2"]}'
```